core/src/test/java/com/radixdlt/api/core/TransactionAccuTreeTest.java

```diff
@@ -0,0 +1,134 @@
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.api.core;
+
+import static com.radixdlt.harness.predicates.NodesPredicate.nodeAt;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.hash.HashCode;
+import com.radixdlt.api.DeterministicCoreApiTestBase;
+import com.radixdlt.crypto.HashUtils;
+import com.radixdlt.harness.predicates.NodePredicate;
+import com.radixdlt.lang.Option;
+import com.radixdlt.transaction.REv2TransactionAndProofStore;
+import java.util.stream.LongStream;
+import org.junit.Test;
+
+public class TransactionAccuTreeTest extends DeterministicCoreApiTestBase {
+
+  @Test
+  public void stateManagerMaintainsCorrectTransactionMerkleTree() {
+    try (var test = buildRunningServerTest(6)) {
+      // Run the setup until 2 epoch proofs are captured
+      var reader = test.getInstance(0, REv2TransactionAndProofStore.class);
+      test.runUntilState(nodeAt(0, NodePredicate.atOrOverEpoch(1)), 1000);
+      var genesisHeader = reader.getEpochProof(1).get().ledgerHeader();
+      test.runUntilState(nodeAt(0, NodePredicate.atOrOverEpoch(2)), 1000);
+      var epochHeader = reader.getEpochProof(2).get().ledgerHeader();
+
+      // Capture the transactions between these 2 proofs
+      var epochTransactions =
+          LongStream.range(
+                  genesisHeader.accumulatorState().stateVersion().toNonNegativeLong().unwrap() + 1,
+                  epochHeader.accumulatorState().stateVersion().toNonNegativeLong().unwrap() + 1)
+              .mapToObj(reader::getTransactionAtStateVersion)
+              .map(Option::unwrap)
+              .toList();
+
+      // Assert a certain count (on which we rely during latter manual merkle computation)
+      assertThat(epochTransactions.size()).isEqualTo(6);
+
+      // Capture the transaction hashes
+      var transactionHashes =
+          epochTransactions.stream()
+              .map(transaction -> HashUtils.blake2b256(transaction.transactionBytes()))
+              .toArray(HashCode[]::new);
+
+      // Assert that header's root hash is equal to manually computed one
+      assertThat(epochHeader.hashes().transactionRoot())
+          .isEqualTo(
+              merkle(
+                  merkle(
+                      merkle(genesisHeader.hashes().transactionRoot(), transactionHashes[0]),
+                      merkle(transactionHashes[1], transactionHashes[2])),
+                  merkle(
+                      merkle(transactionHashes[3], transactionHashes[4]),
+                      merkle(transactionHashes[5], HashUtils.zero256()))));
+
+      // TODO: currently, we cannot test the same for receipt hashes, because the returned receipt
+      // bytes contain all ledger receipt fields, while the state manager only hashes the
+      // "merklizable" part.
+    }
```

Comment:
Do you think it's worth us checking for different scenarios which might result in different sized slices? Or is that checked well by the slice logic tests?

Reply:
I think the unit tests are pretty detailed, starting from here:
https://github.com/radixdlt/babylon-node/blob/5e81ae11414f203e24f46de57ac8bf8bf3da0976/core-rust/state-manager/src/accumulator_tree/test.rs#L102-L104

They include scenarios with slices containing >1 leafs, and even a scenario where 1000 leaves are appended in 1 slice vs in 1000 slices, and result in the same root hash (`incrementally_built_tree_has_same_hash_as_directly_built()`).

In integration tests I am only interested in happy-path.